Batch write imported stuff in importmulti

achow101 opened the pull request:
Instead of writing each item to the wallet database individually, do them in batches so that the import runs faster.

This was tested by importing a ranged descriptor for 10,000 keys.

Current master

```
$ time src/bitcoin-cli -regtest -rpcwallet=importbig importmulti '[{"desc": "sh(wpkh([73111820/44h/1h/0h]tpubDDoT2SgEjaU5rerQpfcRDWPAcwyZ5g7xxHgVAfPwidgPDKVjm89d6jJ8AQotp35Np3m6VaysfUY1C2g68wFqUmraGbzhSsMF9YBuTGxpBaW/1/*))#3w7php47", "range": [0, 10000], "timestamp": "now", "internal": true, "keypool": false, "watchonly": true}]'
...

real	7m45.29s
```

This PR:

```
$ time src/bitcoin-cli -regtest -rpcwallet=importbig4 importmulti '[{"desc": "pkh([73111820/44h/1h/0h]tpubDDoT2SgEjaU5rerQpfcRDWPAcwyZ5g7xxHgVAfPwidgPDKVjm89d6jJ8AQotp35Np3m6VaysfUY1C2g68wFqUmraGbzhSsMF9YBuTGxpBaW/1/*)#v65yjgmc", "range": [0, 10000], "timestamp": "now", "internal": true, "keypool": false, "watchonly": true}]'
...

real	3.93s
```

Fixes #15739

achow101 committed:
Add AddWatchOnlyWithDB, AddKeyOriginWithDB, AddCScriptWithDB functions

AddWatchOnlyWithDB, AddKeyOriginWithDB, and AddCScriptWithDB add their
respective data to the wallet using the provided WalletBatch instead
of creating a new WalletBatch object every time. This allows for batching
writes to the database.

achow101 committed:
Batch writes for importmulti

When writing all of the imported data to the wallet, use a common
WalletBatch object so that batch writes are done and the writes
finish more quickly.

AddKeypoolPubkey is no longer needed so it is also removed

gwillen commented:
utACK although it would be good to factor out the counting into a WalletWriteCache or something, that flushes itself at intervals so you don't have to do it manually.

gwillen commented on src/wallet/rpcdump.cpp:
```diff
@@ -1282,33 +1287,46 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
              const CPubKey& pubkey = entry->second;
              CPubKey temp;
-             if (!pwallet->GetPubKey(id, temp) && !pwallet->AddWatchOnly(GetScriptForRawPubKey(pubkey), timestamp)) {
+             if (!pwallet->GetPubKey(id, temp) && !pwallet->AddWatchOnlyWithDB(*batch, GetScriptForRawPubKey(pubkey), timestamp)) {
```
There is a weird indentation glitch here, although it looks like it was preexisting.

Empact replied:
What are you referring to? I don't see it.

Empact replied:
Ah, think I see it now - 1283-1287 are off by one.

gwillen commented on src/wallet/rpcdump.cpp:
```diff
@@ -1282,33 +1287,46 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
                 pwallet->SetAddressBook(dest, label, "receive");
             }
         }
+        batch.reset();
```
Is this necessary? It's about to go out of scope, right?

achow101 replied:
Probably not. I did that just as a belt-and-suspenders thing.

Empact committed:
Move some of ProcessImport into CWallet::Import*

This maintains encapsulation of CWallet::database in the face of
batching, e.g. allows making the `WithDB` methods private.

Empact commented:
https://github.com/Empact/bitcoin/commit/dd5b0aaf58d843c2532b3ca1829bf90482e9dc52 `CWallet::AddKeyOrigin(const CPubKey& pubkey, const KeyOriginInfo& info)` and `CWallet::WriteKeyMetadata(const CKeyMetadata& meta, const CPubKey& pubkey, const bool overwrite)` can be removed.

https://github.com/Empact/bitcoin/commit/8896384a6c04eb75dcd01742adfb84bde76e9e46 How about moving this functionality into CWallet, so that we don't operate externally on CWallet::database any more than we already do. As the `GetDBHandle` comments say: "Ideally this function would not be necessary.". Practically speaking, this also explicitly limits the scope of the batch operations on `database` to within the lifecycle of the `CWallet` instance.

Empact committed:
Apply the batch treatment to CWallet::SetAddressBook via ImportScriptPubKeys

Empact commented on src/wallet/wallet.h:
```diff
@@ -1222,6 +1224,7 @@ class CWallet final : public CCryptoKeyStore, private interfaces::Chain::Notific
 
     /** Add a KeyOriginInfo to the wallet */
     bool AddKeyOrigin(const CPubKey& pubkey, const KeyOriginInfo& info);
+    bool AddKeyOriginWithDB(WalletBatch& batch, const CPubKey& pubkey, const KeyOriginInfo& info);
```
nit: May be a bad example to follow, but `AddKeypoolPubkeyWithDB` has `batch` as the last argument, which makes as much sense to me.

achow101 replied:
I was following the `AddKeyPubKeyWithDB` way of argument ordering.

Empact commented:
Concept ACK

promag reviewed (commented):
Concept ACK, nice improvement.

MarcoFalke commented:
See also:  

* Use a single wallet batch for UpgradeKeyMetadata #15433

achow101 commented:
I've pulled in @Empact's changes (Empact@dd5b0aa was squashed into 59ec41257f37f2a0438b10ef9776bf7ea3488e9a, cherry-picked the other two commits).

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -1263,17 +1263,22 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
                  throw JSONRPCError(RPC_WALLET_ERROR, "Error adding key to wallet");
              }
              pwallet->UpdateTimeFirstKey(timestamp);
+             if (++cnt % 1000 == 0) {
```
The batch can get very large in memory if lots of keys are involved, so we limit the size to avoid blowing up memory during imports.

gwillen commented:
Especially if there are now multiple places we're manually batching writes like this, and more where we're considering it: Let me suggest again and more strongly that the batch object should be handling autoflushing after every X calls, instead of open-coding it in the caller like this. If it's appropriate for every caller, it could go into WalletBatch, else into a new class that's an extremely thin wrapper around it. I bet the total number of lines added in this PR would actually go down.

Sjors commented:
Concept ACK. I tried the first two commits (so without address book change) with #14145 and importing 2x 1000 keys seems about 6 times faster.

Empact commented:
@gwillen Agree it should be done but would be appropriate as a follow-up

practicalswift commented:
Looks like `AddKeypoolPubkey` is no longer used and can be removed?

achow101 committed:
Have WalletBatch automatically flush every 1000 updates

Since it now automatically flushes, we don't need to have
UpgradeKeyMetadata count and flush separately

achow101 commented:
I've implemented @gwillen's suggestion. `WalletBatch` will flush to disk automatically after every 1000 writes. This appears to be faster too as it does not close and reopen the database on every flush. With this change, I've also updated `UpgradeKeyMetadata`.

> Looks like `AddKeypoolPubkey` is no longer used and can be removed?

Done

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -389,15 +388,10 @@ void CWallet::UpgradeKeyMetadata()
             }
         }
     }
-    batch.reset(); //write before setting the flag
```
in commit ebb0ea85539b3e2f515fca89231c20ac86fb52d5:

Why is this no longer needed?

achow101 replied:
When batch goes out of scope, it will write. This was just a belt and suspenders.

gwillen replied:
The comment suggests that in this case -- unlike the batch write in importmulti -- there's a specific reason to force it to write before the end of the scope.

achow101 replied:
reverted

achow101 commented:
I'm unsure why tests are currently failing.

promag commented on src/wallet/walletdb.h:
```diff
@@ -157,6 +157,9 @@ class WalletBatch
             return false;
         }
         m_database.IncrementUpdateCounter();
+        if (m_database.nUpdateCounter % 1000 == 0) {
```
I get that this makes the code simple, but IMO this defeats the batch purpose. Like, it shouldn't be something to worry about here.

gwillen replied:
I don't understand what you mean by 'defeats the batch purpose'. Are there cases where the caller is depending on the batch not being flushed?

I think it is far better engineering practice to deal with periodic flushing once, inside the batch-management code, than to deal with it independently at each callsite. If there's some situation where periodic flushing is undesirable, I think it would be better to offer both versions.

promag replied:
@gwillen I mean you should be able to create a batch with 1001 changes no? I was suggesting something on top of `WalletBatch` to do that.

gwillen replied:
Well, you can add as many changes as you want, it just flushes after every 1000. This should be completely transparent to the user. Unless there is some reason that flushing can hurt, like the caller is using the batch as though it were an atomic transaction?

achow101 replied:
> Unless there is some reason that flushing can hurt, like the caller is using the batch as though it were an atomic transaction?

There is only one place where an atomic transaction is used, and if an atomic transaction is active, then `Flush()` will do nothing.

Empact replied:
How about add an optional argument to `WalletBatch` that indicates the # of updates to flush after? This makes the operation opt-in/explicit and also allows for optionally adjusting the flushes to the number of operations - e.g. 2x600 rather than 1000 and 200, if that's desirable.

promag replied:
Well I may be wrong but I see the batch as an atomic transaction. Making 1500 updates and then aborting the batch should not commit 1000 updates IMO.

achow101 replied:
> Well I may be wrong but I see the batch as an atomic transaction. Making 1500 updates and then aborting the batch should not commit 1000 updates IMO.

Except that's not how the batch works normally. Batches are not atomic unless you explicitly make them so by starting a transaction with the underlying `BerkeleyBatch`.  In that case, then the flushing won't do anything. In the normal case, each update in a batch itself is atomic and should be expected to be written to the wallet file. There is no aborting. As soon as a batch goes out of scope, every single update is committed to the file. If you make 1500 updates, the first 1000 will be written after you do the 1000th write operation, and then the last 500 will be written when the batch object is deleted.

achow101 replied:
I believe that is out of date

achow101 replied:
I've updated the comment.

promag commented on src/wallet/db.cpp:
```diff
@@ -614,7 +614,9 @@ void BerkeleyBatch::Flush()
         nMinutes = 1;
 
-    env->dbenv->txn_checkpoint(nMinutes ? gArgs.GetArg("-dblogsize", DEFAULT_WALLET_DBLOGSIZE) * 1024 : 0, nMinutes, 0);
+    if (env) {
```
Why this change?

achow101 replied:
So that tests pass. When there's a dummy database, env will be `nulltpr` and this causes a segfault in tests which use a dummy database.

promag replied:
Can't we make `env` is defined instead? Otherwise a comment would make sense as this doesn't really happen-

achow101 replied:
> Can't we make `env` is defined instead?

No. `env` is nullptr for dummy databases which are used in tests. A check like this is used in many other places in this file in the form of `BerkeleyDatabase::IsDummy()`.

> Otherwise a comment would make sense as this doesn't really happen.

Done.

Empact replied:
nit: could deal with it earlier, and use `IsDummy` to make it more explicit.

achow101 replied:
`IsDummy` is done by `BerkeleyDatabase`, not `BerkeleyBatch` so is not available to do here (although the check is completely identical).

Empact commented on src/wallet/wallet.cpp:
```diff
@@ -390,10 +389,6 @@ void CWallet::UpgradeKeyMetadata()
-                if (++cnt % 1000 == 0) {
-                    // avoid creating overlarge in-memory batches in case the wallet contains large amounts of keys
-                    batch.reset(new WalletBatch(*database));
-                }
```
nit: could squash this into the previous

achow101 replied:
Done

Empact commented on src/wallet/wallet.cpp:
```diff
@@ -3404,13 +3404,6 @@ bool CWallet::TopUpKeyPool(unsigned int kpSize)
-    AddKeypoolPubkeyWithDB(pubkey, internal, batch);
-    NotifyCanGetAddressesChanged();
-}
-
```
nit: could squash into https://github.com/bitcoin/bitcoin/pull/15741/commits/c088227bd612920f0e2ea3a34d6e5037490d62c2

achow101 replied:
Done

Sjors reviewed (changes requested):
tACK 396d065 modulo notification bug (inline)

I'm particularly happy that this moves some complexity out of the RPC codebase.

Importing 1000 pubkeys on an iMac with 3,4 GHz Intel Core i5 goes from 25 to 6 seconds.

promag committed:
wallet: Pass WalletBatch to CWallet::UnsetWalletFlag

promag commented:
@achow101 looks like UnsetWalletFlag should reuse the batch instance:
<img width="670" alt="Screenshot 2019-05-18 at 10 45 19" src="https://user-images.githubusercontent.com/3534524/57967966-95483400-795c-11e9-817f-07c3174a05a9.png">
Please test/banchmark with 76c3906d751442ddb1b9bf1662d3e2effb53e950 (https://github.com/promag/bitcoin/tree/pr15741), at least here it gives a significant improvement.

jonasschnelli commented:
Concept ACK

jb55 commented:
@promag's patch brings my import speeds down from 8 minutes to about 10 seconds

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -1288,20 +1289,20 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
             // Add to keypool only works with pubkeys
             if (add_keypool) {
-                pwallet->AddKeypoolPubkey(pubkey, internal);
+                pwallet->AddKeypoolPubkeyWithDB(pubkey, internal, batch);
```
Done

achow101 commented:
I've pulled in @promag's patch.

promag commented:
@achow101 please update times in OP.

achow101 commented:
I noticed that importing scripts would also be slow so I've added `AddCScriptWIthDB` in order to apply the same batching performance boost to such imports.

Also updated OP.

jb55 commented:
tACK 87f7befca04a095a06af003f9fe20d2e3c155d2f sooo fast. tested a rescan with my all my trezor keys from HWI and it worked great. thanks @achow101, hw wallet + core is actually usable now!

promag commented:
Restarted travis, it was failing with 😕
```
47/121 - rpc_psbt.py failed, Duration: 5 s
stdout:
2019-05-18T17:46:43.670000Z TestFramework (INFO): Initializing test directory /tmp/test_runner_₿_🏃_20190518_173555/rpc_psbt_71
2019-05-18T17:46:47.538000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/test_framework.py", line 194, in main
    self.run_test()
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/rpc_psbt.py", line 156, in run_test
    assert_raises_rpc_error(-22, "", self.nodes[0].converttopsbt, signedtx['hex'])
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/util.py", line 105, in assert_raises_rpc_error
    assert try_rpc(code, message, fun, *args, **kwds), "No exception raised"
AssertionError: No exception raised
```

promag commented:
utACK 87f7bef. nit, ampersand should be before space.

Empact commented on src/wallet/wallet.h:
```diff
@@ -1204,6 +1204,7 @@ class CWallet final : public CCryptoKeyStore, private interfaces::Chain::Notific
 
     /** Unsets a single wallet flag */
     void UnsetWalletFlag(uint64_t flag);
+    void UnsetWalletFlagWithDB(WalletBatch& batch, uint64_t flag);
```
nit: could be `private`

achow101 replied:
Done

Empact commented on src/wallet/wallet.h:
```diff
@@ -872,6 +883,7 @@ class CWallet final : public CCryptoKeyStore, private interfaces::Chain::Notific
     //! Adds an encrypted key to the store, without saving it to disk (used by LoadWallet)
     bool LoadCryptedKey(const CPubKey &vchPubKey, const std::vector<unsigned char> &vchCryptedSecret);
     bool AddCScript(const CScript& redeemScript) override;
+    bool AddCScriptWithDB(WalletBatch& batch, const CScript& script);
```
nit: could be `private`

achow101 replied:
Done

Empact commented:
utACK https://github.com/bitcoin/bitcoin/pull/15741/commits/87f7befca04a095a06af003f9fe20d2e3c155d2f

jb55 commented:
utACK 0db94e5

ariard commented:
Tested ACK 0db94e5

Drop import time for 10000 keys ranged descriptor from 495.57s to 3.626s with 2.8 GHz Intel Core i7.

Empact commented:
re-utACK https://github.com/bitcoin/bitcoin/pull/15741/commits/0db94e55dcbbfc741df463c404818d9033b4fff1 only change is re the privacy of `UnsetWalletFlagWithDB` and `AddCScriptWithDB`.
